test/test_closure_pte.cpp

```diff
@@ -0,0 +1,134 @@
+    constexpr Real Q = 4.115 * MJ_per_kg;
+    eos_arr[2] = DavisProducts(a, b, k, n, vc, pc, Cv).Modify<ShiftedEOS>(-Q);
+
+    // Move EOS array from host to device
+    for (auto eos : eos_arr) {
+      eos.GetOnDevice();
+    }
+    constexpr size_t EOS_bytes = num_eos * sizeof(EOS);
+    EOS *v_EOS = (EOS *)PORTABLE_MALLOC(EOS_bytes);
+    portableCopyToDevice(v_EOS, eos_arr.data(), EOS_bytes);
+
+    GIVEN("A difficult state for the density-temperature PTE solver") {
+      // Define the state
+      constexpr Real spvol_bulk = 6.256037280402093e-01;
+      constexpr Real sie_bulk = -1.441692060406520e+10;
+      const std::array<const Real, num_eos> mass_frac = {
+          1.10382442033331e-10, 0.124935312146569, 0.875064687743048};
+
+      // Initial guess: all materials at the cell density
+      const std::array<const Real, num_eos> vol_frac = mass_frac;
+
+      // Calculate material densities (and corresponding energies) and the total
+      // volume fraction
+      std::array<Real, num_eos> densities;
+      std::array<Real, num_eos> sies;
+      Real vfrac_sum = 0;
+      for (auto i = 0; i < num_eos; ++i) {
+        densities[i] = mass_frac[i] / spvol_bulk / vol_frac[i];
+        sies[i] = sie_bulk * mass_frac[i];
+        vfrac_sum += vol_frac[i];
+      }
+
+      // Initialize pressure and temperature arrays to zero
+      std::array<Real, num_eos> temperatures = {0., 0., 0.};
+      std::array<Real, num_eos> pressures = {0., 0., 0.};
+
+      // Copy values to device (when available)
+      constexpr size_t bytes = num_eos * sizeof(Real);
+      Real *v_densities = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_densities, densities.data(), bytes);
+      Real *v_vol_frac = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_vol_frac, vol_frac.data(), bytes);
+      Real *v_sies = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_sies, sies.data(), bytes);
+      Real *v_temperatures = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_temperatures, temperatures.data(), bytes);
+      Real *v_pressures = (Real *)PORTABLE_MALLOC(bytes);
+      portableCopyToDevice(v_pressures, pressures.data(), bytes);
+
+      THEN("The PTE solver should converge") {
+        // Allocate scratch space for the PTE solver
+        int pte_solver_scratch_size = PTESolverRhoTRequiredScratch(num_eos);
+        double *scratch = (double *)PORTABLE_MALLOC(pte_solver_scratch_size);
+        double **lambdas = (double **)PORTABLE_MALLOC(MAX_NUM_LAMBDAS * num_eos);
+        PTESolverRhoT<decltype(v_EOS), Real *, Real **> method(
+            num_eos, v_EOS, vfrac_sum, sie_bulk, v_densities, v_vol_frac, v_sies,
+            v_temperatures, v_pressures, lambdas, scratch);
+
+        // Solve the PTE system and ensure it converged
+        bool pte_converged = PTESolver(method);
+        CHECK(pte_converged);
+        PORTABLE_FREE(lambdas);
+      }
+    }
+    // Free EOS memory
+    PORTABLE_FREE(v_EOS);
```

Comment:
It looks like you're not calling `Finalize` for the EOS models? Or `PORTABLE_FREE` for `scratch?

Reply:
`PORTABLE_FREE` for scratch caused an error that prevents the test from completing for some reason. Good catch on calling `Finalize()` though. Maybe that's part of it...

Reply:
Added the appropriate `free` statements and realized my mistake: I wasn't allocating enough memory.